drivers/hdc1000: remodeled driver + SAUL integration

haukepetersen opened the pull request:
This PR implements the SAUL interface for the HDC1000 sensor and adds means for auto-initializing it.

haukepetersen commented:
Info: no need to review this PR, I am working on an enhancement for SAUL+auto_init and so this PR will change a bit soon...

jfischer-phytec-iot commented:
@haukepetersen about SAUL+auto_init, I will refactor #4155, let's talk about it :-)

haukepetersen commented:
sure! Though I think the problem at hand is two-fold:
1. optimizing the way auto-init works -> #4155
2. adding means for each device driver to initialize and allocate devices using some kind of `void xxx_auto_init(void)` function, which in terms can be called by auto-init/someone else -> #4658

jfischer-phytec-iot commented:
>2. adding means for each device driver to initialize and allocate devices using some kind of void xxx_auto_init(void) function, which in terms can be called by auto-init/someone else -> #4658

You  do not need #4658,  please review #4155. e.g. sensor driver (i2c) should register self by i2c bus driver (i2c subsystem). i2c bus driver will then initialize all registered sensor drivers.

haukepetersen commented:
> You do not need #4658, please review #4155. e.g. sensor driver (i2c) should register self by i2c bus driver (i2c subsystem). i2c bus driver will then initialize all registered sensor drivers.

Not sure I understand #4155 completely, yet. Seems to me, that there are still some things that are not covered: 
- where and who allocates memory for the device descriptors
- if I register a driver at some sub-system, it still needs a parameterless init funciton?!

jfischer-phytec-iot commented:
>where and who allocates memory for the device descriptors
if I register a driver at some sub-system, it still needs a parameterless init funciton?!

@haukepetersen please look at 3dc0206 and refactored and rebased #4155

haukepetersen commented:
Hej, I won't get to this anytime soon, but will get back to this topic once I am back from my leave...

PeterKietzmann commented:
@jfischer-phytec-iot please rebase.

PeterKietzmann commented:
Whoops. Totally overseen that Hauke is the author so forget my previous comment @jfischer-phytec-iot

kYc0o commented:
Feature freeze -> postponed.

PeterKietzmann commented:
@haukepetersen still WIP? Any future plans?

haukepetersen commented:
Now its not WIP anymore, and I went some steps further:
- cleaned up interface
- made read functions return physical values
- made resolution configurable at initialization time
- added default parameter configuration file
- added SAUL support for the driver

Now all it needs is review and testing (as my only `phyNODE` board is currently bricked...).

PeterKietzmann commented on drivers/include/hdc1000.h:
```diff
@@ -9,126 +10,137 @@
+ */
 #ifndef HDC1000_CONVERSION_TIME
-#define HDC1000_CONVERSION_TIME       26000 /**< Default Conversion Time */
+#define HDC1000_CONVERSION_TIME       (26000)
```
In the implementation you mentioned a needed delay of 6.50ms

haukepetersen replied:
Hoped the comment above tells this: 6.5ms per value (temp + hum) -> 13ms * 2 -> 26ms.

PeterKietzmann replied:
Ok, "for each value" makes sense but why times two? Because 14 Bit needs two bytes and that is a "value"? Well, I don't want to block because of that value. The test should get priority (adaption or deleting)

PeterKietzmann commented on drivers/hdc1000/include/hdc1000_params.h:
```diff
@@ -0,0 +1,75 @@
+    {
+        .name = "hdc1000",
+    },
+};
```
I guess we (/I) confused something about the location of the saul info type. For the bmp180 and si70xx driver it can be found under the apropriate sys/auto_init/saul file.

haukepetersen replied:
Japp. But putting these parameters into the params file makes the most sense: here we define how many devices there are gonna be configured, and so we can here also define different names for different instances... I will prepare a PR for the driver configurations you mention above so we have the same style of mapping for all drivers. Will also look at the device driver guide again and make this more explicit.

haukepetersen replied:
-> see #6385

haukepetersen commented:
- adapted test application
- added asserts to driver

haukepetersen commented:
@PeterKietzmann would you mind to give this one more test-run? Thx!

PeterKietzmann commented:
@haukepetersen I think there is not need for all commits. Would you please squash? ACK for the changes. Tested on a phytec board

haukepetersen commented:
actually no :-) I logically grouped the changes to each module in RIOT into one commit each, like I am doing for the last 3 years... So far this seemed to be the way to go?!

haukepetersen commented:
thanks :-)

immesys commented:
Wait, I know I am late to the party, but why does this driver now ignore all errors from I2C? In real life, spurious I2C errors occur with some low probability and they absolutely must be handled. This code doesn't check if i2c_read_bytes even worked!!

immesys commented:
Also, can someone explain how the application calls get_results using the auto_initialized device handle? How do I get the handle from the application?
